solo-io/gloo
# Glooctl grpc Upstream output in augmented status

# Description

In CLI code, augment the namespaced statuses for Upstreams to contain gRPC function names, if applicable, before passing them to `cliutils.PrintList()` so that they are present in glooctl output

## Code changes
- Add `addFunctionsFromGrpcTranscoder()` helper function that returns a function which adds gRPC function names to a given Upstream's namespaced status, if applicable, for the namespace passed to the parent function
  - Add various sub-helper functions
  - Unit test 
- `PrintUpstreams()` takes a `namespace` argument and invokes `upstreams.Each(addFunctionsFromGrpcTranscoder(namespace))`
- Various cleanup/comments/etc.

# Context

See [design doc](https://soloio.slab.com/posts/glooctl-upstream-g-rpc-functions-output-ieg6qc6b) and [slack channel](https://solo-io-corp.slack.com/archives/C07CBV7EFC6)

Note that the customer requesting this uses jq in an automated process to retrieve function names
An example of how this can be done is below in the Testing Steps section

## Interesting decisions
 
We augment the status in CLI code rather than controller code (which would lead the augmented status to be written to etcd) in order to avoid increasing the size of written resources

We opt only to handle the in-line descriptor bin case, and not descriptors referenced in files or configmaps, as this handles Upstreams that are generated by Discovery
#9798 has been opened as a followup to handle those additional cases
 
See design doc and slack channel for additional context on alternate solutions that were considered

## Testing steps

- Follow the [Gloo docs](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/grpc/simple-routing/) for setting up a basic gRPC service
- Use current glooctl to get the Upstream and observe that the function names are not present (try both JSON and YAML; YAML shown here):
```
$ glooctl get upstream default-grpcstore-demo-80 -o yaml

---
discoveryMetadata:
  labels:
    app: grpcstore-demo
kube:
  selector:
    app: grpcstore-demo
  serviceName: grpcstore-demo
  serviceNamespace: default
  servicePort: 80
  serviceSpec:
    grpcJsonTranscoder:
      matchIncomingRequestRoute: true
      protoDescriptorBin: Cqw...zM=
      services:
      - solo.examples.v1.StoreService
metadata:
  generation: "4"
  labels:
    discovered_by: kubernetesplugin
    discovery.solo.io/function_discovery: enabled
  name: default-grpcstore-demo-80
  namespace: gloo-system
  ownerReferences: []
  resourceVersion: "1035358"
namespacedStatuses:
  statuses:
    gloo-system:
      reportedBy: gloo
      state: Accepted
```
- Build the CLI locally using `make build-cli-local`
- Invoke the locally built CLI and observe the augmented output:
```
$ ./_output/glooctl-darwin-amd64 get upstream default-grpcstore-demo-80 -o yaml

---
discoveryMetadata:
  labels:
    app: grpcstore-demo
kube:
  selector:
    app: grpcstore-demo
  serviceName: grpcstore-demo
  serviceNamespace: default
  servicePort: 80
  serviceSpec:
    grpcJsonTranscoder:
      matchIncomingRequestRoute: true
      protoDescriptorBin: Cqw...zM=
      services:
      - solo.examples.v1.StoreService
metadata:
  generation: "4"
  labels:
    discovered_by: kubernetesplugin
    discovery.solo.io/function_discovery: enabled
  name: default-grpcstore-demo-80
  namespace: gloo-system
  ownerReferences: []
  resourceVersion: "1035358"
namespacedStatuses:
  statuses:
    gloo-system:
      details:
        fields:
          functionNames:
            structValue:
              fields:
                solo.examples.v1.StoreService:
                  listValue:
                    values:
                    - stringValue: CreateItem
                    - stringValue: ListItems
                    - stringValue: DeleteItem
                    - stringValue: GetItem
      reportedBy: gloo
      state: Accepted
```
- Note that we can get the function names in a more concise format using jq:
```
./_output/glooctl-darwin-amd64 get upstream default-grpcstore-demo-80 -o json | jq '.[].namespacedStatuses.statuses.[].details.fields.functionNames.structValue.fields | to_entries | map({(.key): [.value.listValue.values.[].stringValue]})'

[
  {
    "solo.examples.v1.StoreService": [
      "CreateItem",
      "ListItems",
      "DeleteItem",
      "GetItem"
    ]
  }
]
```

## Notes for reviewers

This will be backported to 1.14, the first version that deprecated the `Grpc` ServiceSpec API

I'm not sure what, if any docs, it would be worth to update with regards to this change
Perhaps @solo-io/solo-docs could weigh in on that point

# Checklist:
